File: Predict_the_Winner.cpp
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        int n = nums.size();
        vector<int> max_diff(n, 0);

        for (int i = n - 1; i >= 0; i--) {
            max_diff[i] = nums[i];
            for (int j = i + 1; j < n; j++) {
                max_diff[j] = max(nums[i] - max_diff[j], nums[j] - max_diff[j - 1]);
            }
        }

        return max_diff[n - 1] >= 0;        
    }
};

File: Predict_the_Winner.py
class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        n = len(nums)
        max_diff = [0] * n

        for i in range(n - 1, -1, -1):
            max_diff[i] = nums[i]
            for j in range(i + 1, n):
                max_diff[j] = max(nums[i] - max_diff[j], nums[j] - max_diff[j - 1])
        
        return max_diff[n - 1] >= 0
